### PR TITLE
Add FBC applications & components

### PR DESCRIPTION
**Pull request by creydr**
Adding the applications and components for the FBCs

### CODE REVIEW

File: pkg/prowgen/prowgen_konflux.go [hunk 1]
```diff
@@ -341,6 +345,44 @@ func GenerateKonfluxServerlessOperator(ctx context.Context, openshiftRelease Rep
...
 
+func generateFBCApplications(soMetadata *project.Metadata, openshiftRelease Repository, r Repository, branch string, release string, resourceOutputPath string, buildArgs []string) error {
+	soMetadata.Requirements = &project.Requirements{
+		OcpVersion: project.OcpVersion{
+			List: []string{
+				"4.13",
+				"4.17",
+			},
+		},
+	}
```
Comment (creydr) [inline pkg/prowgen/prowgen_konflux.go:356]: This is only temporary for testing until https://github.com/openshift-knative/serverless-operator/pull/2926/ is in

File: pkg/konfluxgen/konfluxgen.go [hunk 2]
```diff
@@ -428,6 +447,149 @@ func Generate(cfg Config) error {
...
+	}
+
+	// Generate application yaml
+	buf := &bytes.Buffer{}
+	if err := applicationTemplate.Execute(buf, config); err != nil {
+		return fmt.Errorf("failed to execute template for application %q: %w", appKey, err)
+	}
+	if err := os.WriteFile(appPath, buf.Bytes(), 0777); err != nil {
+		return fmt.Errorf("failed to write application file %q: %w", appPath, err)
+	}
```
Comment (creydr) [inline pkg/konfluxgen/konfluxgen.go:524]: Duplications from [konfluxgen.Generate()](https://github.com/openshift-knative/hack/blob/f2a9e86ff33b53185aabec1a55f84660f3523a15/pkg/konfluxgen/konfluxgen.go#L326). But before refactoring it, I want to align on the general approach first

File: pkg/konfluxgen/konfluxgen.go [hunk 3]
```diff
@@ -428,6 +447,149 @@ func Generate(cfg Config) error {
...
+		ProjectDirectoryImageBuildStepConfiguration: cioperatorapi.ProjectDirectoryImageBuildStepConfiguration{
+			To: cioperatorapi.PipelineImageStreamTagReference(componentKey),
+			ProjectDirectoryImageBuildInputs: cioperatorapi.ProjectDirectoryImageBuildInputs{
+				DockerfilePath: cfg.DockerfilePath,
+				ContextDir:     cfg.ContextDirPath,
+			},
+		},
+		Pipeline:                      FBCBuild,
+		Tags:                          append(cfg.Tags, "latest"),
+		BuildArgs:                     cfg.BuildArgs,
```
Comment (creydr) [inline pkg/konfluxgen/konfluxgen.go:512]: need to add the OLM_IMAGE arg override too

File: pkg/konfluxgen/konfluxgen.go (hunk 2 above)
Comment (pierDipi) [reply pkg/konfluxgen/konfluxgen.go:524]: Agree there is a lot of duplication

Comment (pierDipi) [reply pkg/konfluxgen/konfluxgen.go:524]: we could "augment" `collectConfigurations` with an optional function for the additional images/components

File: pkg/konfluxgen/konfluxgen.go (hunk 3 above)
Comment (creydr) [reply pkg/konfluxgen/konfluxgen.go]: done

File: pkg/konfluxgen/konfluxgen.go (hunk 2 above)
Comment (creydr) [reply pkg/konfluxgen/konfluxgen.go]: :+1: 
Updated

File: pkg/prowgen/prowgen_konflux.go [hunk 4]
```diff
@@ -342,10 +347,97 @@ func GenerateKonfluxServerlessOperator(ctx context.Context, openshiftRelease Rep
...
+						Metadata: cioperatorapi.Metadata{
+							Org:    r.Org,
+							Repo:   r.Repo,
+							Branch: branch,
+						},
+						Images: []cioperatorapi.ProjectDirectoryImageBuildStepConfiguration{
+							{
+								To: cioperatorapi.PipelineImageStreamTagReference(fmt.Sprintf("serverless-index-%s-fbc-%s", release, v)),
+								ProjectDirectoryImageBuildInputs: cioperatorapi.ProjectDirectoryImageBuildInputs{
+									DockerfilePath: "Dockerfile",
```
Comment (creydr) [inline pkg/prowgen/prowgen_konflux.go:383]: This results for example in the following:
```
apiVersion: appstudio.redhat.com/v1alpha1
kind: Component
metadata:
  annotations:
    build.appstudio.openshift.io/pipeline: '{"name":"docker-build","bundle":"latest"}'
  name: serverless-index-135-fbc-413
spec:
  componentName: serverless-index-135-fbc-413
  application: serverless-operator-135-fbc-413
  source:
    git:
      url: https://github.com/openshift-knative/serverless-operator.git
      context: ./olm-catalog/serverless-operator-index/v4.13
      dockerfileUrl: Dockerfile
      revision: main
```

From the DockerfilePath doc:

> DockerfilePath is the path to a Dockerfile in the project to run **relative to the context_dir**.

File: pkg/prowgen/prowgen_konflux.go [hunk 5]
```diff
@@ -342,10 +347,97 @@ func GenerateKonfluxServerlessOperator(ctx context.Context, openshiftRelease Rep
...
-func serverlessIndexNudge(downstreamVersion string) string {
-	return konfluxgen.Truncate(konfluxgen.Sanitize(fmt.Sprintf("%s-%s", "serverless-index", downstreamVersion)))
+func serverlessIndexNudges(downstreamVersion string, ocpVersions []string) []string {
+	indexes := []string{
+		konfluxgen.Truncate(konfluxgen.Sanitize(fmt.Sprintf("%s-%s", "serverless-index", downstreamVersion))),
+	}
+
+	for _, v := range ocpVersions {
+		indexes = append(indexes, konfluxgen.Truncate(konfluxgen.Sanitize(fmt.Sprintf("serverless-index-%s-fbc-%s-index", downstreamVersion, v))))
+	}
```
Comment (pierDipi) [inline pkg/prowgen/prowgen_konflux.go:440]: This is just the component name, do you know if it is supposed to work when the application is different?

Comment (pierDipi) [reply pkg/prowgen/prowgen_konflux.go:440]: it seems to be working https://github.com/openshift/hac-dev/pull/909

**Review by pierDipi — APPROVED** [review]
/lgtm
/approve

File: pkg/prowgen/prowgen_konflux.go (hunk 5 above)
Comment (creydr) [reply pkg/prowgen/prowgen_konflux.go:440]: I only saw, that the nudge-ref is only a list of strings (without an application ref). As they are in the same namespace, I'd assume so, but not sure.
Let me confirm with them

Comment (creydr) [reply pkg/prowgen/prowgen_konflux.go:440]: > I only saw, that the nudge-ref is only a list of strings (without an application ref). As they are in the same namespace, I'd assume so, but not sure. Let me confirm with them

Ah didn't see your message https://github.com/openshift-knative/hack/pull/334#discussion_r1799206591

Comment (pierDipi) [reply pkg/prowgen/prowgen_konflux.go:440]: from one test I saw
```
 it('should return all components in a namespace', () => {
    useK8sWatchResourceMock.mockReturnValue([mockComponentsData, true, undefined]);
    const { result } = renderHook(() => useAllComponents('test-ns'));
    const [components] = result.current;
    expect(components).toHaveLength(3);
  });
```
so it should work

**Comment by mgencur** [comment]
Should we also remove this line? https://github.com/openshift-knative/hack/pull/334/files#diff-888ae255f82f46f198de233d2aa02b58b2f97fcc52b07f89ea8cac5d79514ca4R298
The point is to stop generating the Component and Tekton pipeline for FBC in the `serverless-operator-135` application (note the missing OCP version) and only generate it in the OCP specific applications.

**Comment by pierDipi** [comment]
@mgencur I think more that removing that one, we want to add an exclude otherwise will be anyway added as normal component (since it's part of the CI config)

**Comment by creydr** [comment]
https://github.com/openshift-knative/hack/pull/342

**Pull request merged**